67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1096 @@
+					triangles[i].vertexFormat = EF_R32G32B32_SFLOAT;
+					triangles[i].indexType = obj.indexType;
+					triangles[i].geometryFlags = IGPUBottomLevelAccelerationStructure::GEOMETRY_FLAGS::OPAQUE_BIT;
+
+					const auto blasFlags = bitflag(IGPUBottomLevelAccelerationStructure::BUILD_FLAGS::PREFER_FAST_TRACE_BIT) | IGPUBottomLevelAccelerationStructure::BUILD_FLAGS::ALLOW_COMPACTION_BIT;
+
+					blasBuildInfos[i].buildFlags = blasFlags;
+					blasBuildInfos[i].geometryCount = 1;	// only 1 geometry object per blas
+					blasBuildInfos[i].srcAS = nullptr;
+					blasBuildInfos[i].dstAS = nullptr;
+					blasBuildInfos[i].triangles = &triangles[i];
+					blasBuildInfos[i].scratch = {};
+
+					{
+						const uint32_t maxPrimCount[1] = { primitiveCounts[i] };
+						buildSizes[i] = m_device->getAccelerationStructureBuildSizes(blasFlags, false, std::span{&triangles[i], 1}, maxPrimCount);
+					}
+
+					maxScratchSize = std::max(maxScratchSize, buildSizes[i].accelerationStructureSize);
+
+					{
+						IGPUBuffer::SCreationParams params;
+						params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_ACCELERATION_STRUCTURE_STORAGE_BIT;
+						params.size = buildSizes[i].accelerationStructureSize;
+						smart_refctd_ptr<IGPUBuffer> asBuffer = createBuffer(params);
+
+						IGPUBottomLevelAccelerationStructure::SCreationParams blasParams;
+						blasParams.bufferRange.buffer = asBuffer;
+						blasParams.bufferRange.offset = 0u;
+						blasParams.bufferRange.size = buildSizes[i].accelerationStructureSize;
+						blasParams.flags = IGPUBottomLevelAccelerationStructure::SCreationParams::FLAGS::NONE;
+						gpuBlas[i] = m_device->createBottomLevelAccelerationStructure(std::move(blasParams));
+					}
+				}
+
+				auto cmdbufBlas = getSingleUseCommandBufferAndBegin(pool);
+				cmdbufBlas->beginDebugMarker("Build BLAS");
+
+				cmdbufBlas->resetQueryPool(queryPool.get(), 0, objectsGpu.size());
+
+				smart_refctd_ptr<IGPUBuffer> scratchBuffer;
+				{
+					IGPUBuffer::SCreationParams params;
+					params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+					params.size = maxScratchSize;
+					scratchBuffer = createBuffer(params);
+				}
+
+				uint32_t queryCount = 0;
+				for (uint32_t i = 0; i < objectsGpu.size(); i++)
+				{
+					blasBuildInfos[i].dstAS = gpuBlas[i].get();
+					blasBuildInfos[i].scratch.buffer = scratchBuffer;
+					blasBuildInfos[i].scratch.offset = 0u;
+
+					IGPUBottomLevelAccelerationStructure::BuildRangeInfo buildRangeInfos[1u];
+					buildRangeInfos[0].primitiveCount = primitiveCounts[i];
+					buildRangeInfos[0].primitiveByteOffset = 0u;
+					buildRangeInfos[0].firstVertex = 0u;
+					buildRangeInfos[0].transformByteOffset = 0u;
+					IGPUBottomLevelAccelerationStructure::BuildRangeInfo* pRangeInfos[1u];
+					pRangeInfos[0] = &buildRangeInfos[0];
+
+					cmdbufBlas->buildAccelerationStructures({ &blasBuildInfos[i], 1 }, pRangeInfos);
```

Comment:
why areyou building one and barriering one at a time?

Build all blas at once! that way you can hope to saturate the GPU (the builds run in parallel)